### ISSUE TITLE
Serial multiplier example - can't seem to get it working

### ISSUE THREAD

**Issue by georgejkaye** [janestreet/hardcaml#8]
I'm trying out the serial multiplier example in the documentation but I can't seem to get the Hardcaml Bits implementation working. I'm using hardcaml v0.14.1 (the one on OPAM).

```ocaml
# open Hardcaml
# open Hardcaml.Bits
# let rec umul' a b =
        if to_int b = 0 
        then 
            zero (width a)
        else
            let partial_product = mux2 b.:[0,0] a (zero (width a)) in
            partial_product +: umul' (sll a 1) (srl b 1)
val umul' : t -> t -> t = <fun>
# let umul a b = umul' (uresize a (width a + width b)) b
val umul : t -> t -> t = <fun>
# to_int (umul (of_int ~width:2 3) (of_int ~width:3 5));;
- : int = 21
```

The last line be evaluating to 15 (which is 3 * 5), but it is instead evaluating to 21!
The issue seems to be arising from the use of `b.:[0,0]` and `mux2`, where _every_ iteration of the algorithm is being added to the sum, not only those where b mod 2 = 1.
I have tested it with this simple example below:

```ocaml
# let b = of_int ~width:2 2
val b  : t = 10
# b. :[0,0]
- : t = 0
# mux2 b.:[0,0] (of_int ~width:1 1) (of_int ~width:1 0)
- : t = 1
```

and the issue still seems to arise.

Is there something I'm missing here?

**Comment by andrewray** [comment]
The examples are really written for the bleeding edge version of ocaml, which is not yet released and requires some opam hackery.

However, I tried this specific example out on an Ocaml-4.11.1 compiler with Hardcaml v0.14.1 and it seemed to work fine.

What error are you seeing?  Is there anything unusual about the platform you are running on?

**Comment by georgejkaye** [comment]
The error is that 5 * 3 should be equal to 15 (as shown on the example page) but the multiplier is returning 21 :(

(I tested it out on another machine and it also returns 21. Both machines are running ocaml 4.12.0)

**Comment by andrewray** [comment]
Gotcha, and I can repeat the error.

This seems like a bug in `Bits.select`.

```
select (of_string "10") 0 0
- : t = 0
```

which is fine, but:

```
to_int (select (of_string "10") 0 0)
- : int = 2
```

The select operation is supposed to mask the result to the required length, but in this case is copying the extraneous bits.  Here's a hack which should fix things for you.  But realistically, you will need to update to the latest version to really work round this.

```
let select s h l = (select s h l) &: (ones (h-l+1));;
let (.:[]) s (h,l) = select s h l
```

**Comment by andrewray** [comment]
I am going ask around here how hard to would be to release a 0.14.2 hotfix for this branch.

**Comment by georgejkaye** [comment]
Phew, was starting to doubt that 3 * 5 really did equal 15! Thanks for the workaround, it should be enough for me while I'm just exploring the library.

**Comment by andrewray** [comment]
I know what the bug is now - there is a mistake in the C-stubs bindings for bytecode operation.  This would work properly if you compiled into native code.  We only actually caught this problem when we wrote the documentation as this is roughly the only things we compile as bytecode.

I am keen to get the issue patched if possible.

**Comment by andrewray** [comment]
This should now be fixed in v0.14.2 on opam (and FWIW the bleeding edge repos already had this fix - I do recommend them).

I made sure 3*5 = 15, not 21.  Sanity preserved.

Thanks to @cwong-ocaml for pushing out the release ahead of our usual schedule.
